File: Linked_List/024.Swap-Nodes-in-Pairs/024.Swap-Nodes-in-Pairs.ts
function swapPairs(head: ListNode | null): ListNode | null {
  let start = head;
  let end = head;
  for (let i = 0; i < 2; i++) {
    if (end === null) {
      return head;
    }
    end = end.next;
  }
  let newHead = swap(start, end);
  start.next = swapPairs(end);
  return newHead;
}

function swap(p: ListNode, head: ListNode | null) {
  let prev = null;
  let next = p;
  while (p !== head) {
    next = p.next;
    p.next = prev;
    prev = p;
    p = next;
  }
  return prev;
}

File: Linked_List/024.Swap-Nodes-in-Pairs/024.Swap-Nodes-in-Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        dummy = ListNode(0, head)
        current = dummy

        while current and current.next and current.next.next:
            node1 = current.next
            node2 = current.next.next
            postP = current.next.next.next
            node1.next = postP
            node2.next = node1
            current.next = node2

            current = current.next.next
            
            
        
        return dummy.next
